Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift

```diff
@@ -529,45 +529,87 @@ extension BrowserViewController: WKNavigationDelegate {
     return .allow
   }
 
-  // When this function is moved to Async-Await, do NOT forget to move the `TabManagerNavDelegate` to async-await!
-  public func webView(_ webView: WKWebView, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void) {
+  @MainActor
+  public func webView(_ webView: WKWebView, respondTo challenge: URLAuthenticationChallenge) async -> (URLSession.AuthChallengeDisposition, URLCredential?) {
 
     // If this is a certificate challenge, see if the certificate has previously been
     // accepted by the user.
     let origin = "\(challenge.protectionSpace.host):\(challenge.protectionSpace.port)"
     if challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
       let trust = challenge.protectionSpace.serverTrust,
       let cert = SecTrustGetCertificateAtIndex(trust, 0), profile.certStore.containsCertificate(cert, forOrigin: origin) {
-      completionHandler(.useCredential, URLCredential(trust: trust))
-      return
+      return (.useCredential, URLCredential(trust: trust))
+    }
+    
+    // Certificate Pinning
+    if challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust {
+      if let serverTrust = challenge.protectionSpace.serverTrust {
+        let host = challenge.protectionSpace.host
+        let port = challenge.protectionSpace.port
+        
+        return await Task.detached(priority: .userInitiated) {
+          let result = BraveCertificateUtility.verifyTrust(serverTrust,
+                                                           host: host,
+                                                           port: port)
+          let certificateChain = (0..<SecTrustGetCertificateCount(serverTrust))
+            .compactMap { SecTrustGetCertificateAtIndex(serverTrust, $0) }
+          
+          // Cert is valid and should be pinned
+          if result == 0 {
+            return (.useCredential, URLCredential(trust: serverTrust))
+          }
+          
+          // Cert is valid and should not be pinned
+          // Let the system handle it and we'll show an error if the system cannot validate it
+          if result == Int32.min {
+            return (.performDefaultHandling, nil)
+          }
+          
+          return await Task { @MainActor in
+            // Cert is invalid and cannot be pinned
+            Logger.module.error("CERTIFICATE_INVALID")
+            let errorCode = CFNetworkErrors.braveCertificatePinningFailed.rawValue
+            
+            let underlyingError = NSError(domain: kCFErrorDomainCFNetwork as String,
+                                          code: Int(errorCode),
+                                          userInfo: ["_kCFStreamErrorCodeKey": Int(errorCode)])
```

Comment:
Private API key?

Reply:
Yeah. It is used in our certificate pinning. This is the only way known to get certificate error codes :(